File: src/lib/agents/repo-analyzer.ts
// ============================================================
// Repo Analyzer Agent
// Analyzes repos and extracts deep capability signals
// ============================================================

import { RepoInput, CapabilityCategory } from "./types";

const CAPABILITY_KEYWORDS: Record<CapabilityCategory, string[]> = {
  memory: [
    "vector", "embedding", "chroma", "pinecone", "weaviate", "milvus",
    "qdrant", "faiss", "memory", "storage", "redis", "memcached",
    "neo4j", "graph-db", "knowledge-graph", "faiss", "db", "database",
    "sqlite", "postgres", "mongodb", "supabase",
  ],
  agent: [
    "agent", "langchain", "langgraph", "crewai", "autogen", "autogpt",
    "orchestrat", "planning", "multi-agent", "workflow", "task",
    "copilot", "assistant", "bot", "chatbot", "delegate", "swarm",
    "mcp", "tool-use", "function-call",
  ],
  rag: [
    "rag", "retrieval", "search", "semantic", "llamaindex", "llama-index",
    "document", "ingest", "chunk", "embed", "index", "query-engine",
    "hyde", "re-rank", "rerank", "document-load",
  ],
  ui: [
    "react", "vue", "svelte", "nextjs", "next.js", "dashboard", "frontend",
    "component", "ui", "design", "tailwind", "css", "storybook",
    "shadcn", "radix", "chart", "visualization", "web", "app",
  ],
  automation: [
    "automation", "workflow", "pipeline", "n8n", "zapier", "trigger",
    "schedule", "cron", "ci-cd", "github-actions", "webhook",
    "event", "queue", "temporal", "bull", "agenda",
  ],
  "model-serving": [
    "inference", "serving", "deploy", "vllm", "triton", "tensorrt",
    "onnx", "model", "llm", "gpt", "claude", "openai", "huggingface",
    "ollama", "llamacpp", "llama.cpp", "fine-tun", "training",
    "transformer", "diffusion", "stable-diffusion",
  ],
  data: [
    "data", "etl", "pipeline", "stream", "kafka", "spark", "airflow",
    "dbt", "analytics", "metrics", "warehouse", "lake", "parquet",
    "arrow", "pandas", "polars",
  ],
  security: [
    "security", "auth", "oauth", "jwt", "encryption", "privacy",
    "rbac", "firewall", "scan", "vulnerability", "secret",
  ],
  infra: [
    "docker", "kubernetes", "k8s", "terraform", "cloud", "serverless",
    "aws", "gcp", "azure", "helm", "nginx", "caddy", "proxy",
    "monitor", "observ", "log", "tracing",
  ],
  communication: [
    "chat", "real-time", "websocket", "socket", "email", "sms",
    "notification", "slack", "discord", "telegram", "api",
  ],
};

const CAPABILITY_LABELS: Record<CapabilityCategory, string> = {
  memory: "Memory & Storage",
  agent: "Agent Framework",
  rag: "RAG & Retrieval",
  ui: "Frontend & UI",
  automation: "Automation & Workflow",
  "model-serving": "Model Serving & LLM",
  data: "Data & Analytics",
  security: "Security & Auth",
  infra: "Infrastructure & DevOps",
  communication: "Communication & Real-time",
};

const CAPABILITY_ICONS: Record<CapabilityCategory, string> = {
  memory: "Database",
  agent: "Bot",
  rag: "BookOpen",
  ui: "Palette",
  automation: "Workflow",
  "model-serving": "Cpu",
  data: "BarChart3",
  security: "Shield",
  infra: "Server",
  communication: "MessageSquare",
};

export interface AnalyzedRepo extends RepoInput {
  detectedCapabilities: CapabilityCategory[];
  capabilityScores: Record<CapabilityCategory, number>;
  maturityLevel: "experimental" | "growing" | "mature" | "industry-standard";
  roleHints: string[];
}

export function analyzeRepo(repo: RepoInput): AnalyzedRepo {
  const combined = `${repo.name} ${repo.description || ""} ${(repo.topics || []).join(" ")} ${repo.innovationSignals.join(" ")}`.toLowerCase();

  const capabilityScores: Record<CapabilityCategory, number> = {} as any;
  const detectedCapabilities: CapabilityCategory[] = [];

  for (const [cat, keywords] of Object.entries(CAPABILITY_KEYWORDS)) {
    let score = 0;
    for (const kw of keywords) {
      if (combined.includes(kw)) {
        score += kw.includes(repo.language?.toLowerCase() || "") ? 3 : kw.length > 5 ? 2 : 1;
      }
    }
    capabilityScores[cat as CapabilityCategory] = score;
    if (score > 0) detectedCapabilities.push(cat as CapabilityCategory);
  }

  // Determine maturity
  let maturityLevel: AnalyzedRepo["maturityLevel"] = "experimental";
  if (repo.stars > 50000) maturityLevel = "industry-standard";
  else if (repo.stars > 10000) maturityLevel = "mature";
  else if (repo.stars > 2000) maturityLevel = "growing";

  // Generate role hints
  const roleHints: string[] = [];
  if (detectedCapabilities.includes("memory")) roleHints.push("Storage Layer");
  if (detectedCapabilities.includes("agent")) roleHints.push("Orchestration Engine");
  if (detectedCapabilities.includes("rag")) roleHints.push("Knowledge Processor");
  if (detectedCapabilities.includes("ui")) roleHints.push("User Interface");
  if (detectedCapabilities.includes("automation")) roleHints.push("Automation Engine");
  if (detectedCapabilities.includes("model-serving")) roleHints.push("LLM Backend");
  if (detectedCapabilities.includes("data")) roleHints.push("Data Pipeline");
  if (detectedCapabilities.includes("security")) roleHints.push("Security Layer");
  if (detectedCapabilities.includes("infra")) roleHints.push("Infrastructure");
  if (detectedCapabilities.includes("communication")) roleHints.push("Communication Hub");
  if (roleHints.length === 0) roleHints.push("Utility");

  return {
    ...repo,
    detectedCapabilities,
    capabilityScores,
    maturityLevel,
    roleHints,
  };
}

export function analyzeRepos(repos: RepoInput[]): AnalyzedRepo[] {
  return repos.map(analyzeRepo);
}

export { CAPABILITY_LABELS, CAPABILITY_ICONS, CAPABILITY_KEYWORDS };

File: python-backend/agents/repo_analyzer.py
"""
Repo Analyzer — Extracts signals from GitHub repos.
Identifies AI components, API surfaces, complexity levels, and capability hints.
"""

from typing import Any


# Capability detection keywords mapped to categories
CAPABILITY_KEYWORDS: dict[str, list[str]] = {
    "memory": ["vector", "embedding", "storage", "database", "cache", "rag", "retrieval", "semantic", "search"],
    "agent": ["agent", "workflow", "orchestrat", "autonomous", "planning", "decision", "multi-agent", "crew"],
    "rag": ["retrieval", "augmented", "generation", "document", "chunk", "index", "knowledge-base", "embed"],
    "ui": ["ui", "frontend", "react", "vue", "dashboard", "component", "visualization", "chart", "interface"],
    "backend": ["api", "server", "backend", "microservice", "rest", "graphql", "fastapi", "express", "database"],
    "automation": ["automat", "cron", "scheduler", "pipeline", "ci/cd", "deploy", "workflow", "bot", "trigger"],
}


def analyze_repo(repo: dict[str, Any]) -> dict[str, Any]:
    """
    Analyze a single repo to extract signals.

    Returns a dict with:
      - name, description, stars, language
      - signals: { hasAI, hasAPI, complexity, capabilities, entryPoints }
    """
    description = (repo.get("description") or "").lower()
    topics = [t.lower() for t in (repo.get("topics") or [])]
    name = repo.get("name", "").lower()
    stars = repo.get("stars", repo.get("stargazers_count", 0))

    # Detect AI-related signals
    ai_keywords = ["ai", "ml", "machine-learning", "deep-learning", "llm", "gpt", "neural",
                   "transformer", "model", "nlp", "computer-vision", "openai", "claude", "anthropic"]
    has_ai = any(kw in description or kw in name or kw in " ".join(topics) for kw in ai_keywords)

    # Detect API presence
    api_keywords = ["api", "rest", "graphql", "endpoint", "http", "server", "sdk"]
    has_api = any(kw in description or kw in name for kw in api_keywords)

    # Determine complexity
    if stars > 10000:
        complexity = "high"
    elif stars > 1000:
        complexity = "medium"
    else:
        complexity = "low"

    # Map capabilities
    detected_capabilities = []
    all_text = f"{description} {name} {' '.join(topics)}"
    for cap_type, keywords in CAPABILITY_KEYWORDS.items():
        if any(kw in all_text for kw in keywords):
            detected_capabilities.append(cap_type)

    if not detected_capabilities:
        detected_capabilities = ["general"]

    # Infer entry points
    language = repo.get("language", "").lower()
    entry_points = _infer_entry_points(language, description)

    return {
        "name": repo.get("name", repo.get("full_name", "")),
        "full_name": repo.get("full_name", repo.get("name", "")),
        "description": repo.get("description", ""),
        "stars": stars,
        "language": repo.get("language"),
        "topics": topics,
        "signals": {
            "hasAI": has_ai,
            "hasAPI": has_api,
            "complexity": complexity,
            "capabilities": detected_capabilities,
            "entryPoints": entry_points,
        },
    }


def analyze_repos(repos: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Analyze a list of repos and return their signal profiles."""
    return [analyze_repo(repo) for repo in repos]


def _infer_entry_points(language: str, description: str) -> list[str]:
    """Infer likely entry points based on language and description."""
    entry_map = {
        "python": ["main.py", "app.py", "setup.py", "pyproject.toml"],
        "typescript": ["src/index.ts", "src/main.ts", "package.json"],
        "javascript": ["src/index.js", "index.js", "package.json"],
        "go": ["main.go", "cmd/main.go"],
        "rust": ["src/main.rs", "Cargo.toml"],
        "java": ["src/main/java/Main.java", "pom.xml", "build.gradle"],
    }
    desc_lower = description.lower()

    # Check for framework-specific patterns
    if "fastapi" in desc_lower:
        return ["app/main.py", "app/api.py"]
    elif "flask" in desc_lower:
        return ["app.py", "wsgi.py"]
    elif "django" in desc_lower:
        return ["manage.py", "wsgi.py"]
    elif "next.js" in desc_lower or "nextjs" in desc_lower:
        return ["app/page.tsx", "app/layout.tsx"]
    elif "react" in desc_lower:
        return ["src/App.tsx", "src/index.tsx"]

    return entry_map.get(language, ["README.md"])
